src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -0,0 +1,110 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.Monomorphization
+{
+    public class ResolveGenericsSyntax :
+        SyntaxTreeTransformation<ResolveGenericsScope>
+    {
+        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces)
+        {
+            if (namespaces == null || namespaces.Contains(null)) throw new ArgumentNullException(nameof(namespaces));
+
+            // Get list of generics
+            List<QsQualifiedName> generics = new List<QsQualifiedName>();
+            var Callables = namespaces.GlobalCallableResolutions();
+            foreach (var callable in Callables)
+            {
+                if (callable.Value.Signature.TypeParameters.Any())
+                {
+                    generics.Add(callable.Key);
+                }
+            }
+
+            var filter = new ResolveGenericsSyntax(new ResolveGenericsScope());
+            return namespaces.Select(ns => filter.Transform(ns));
+        }
+
+        public static IEnumerable<QsNamespace> Apply(params QsNamespace[] namespaces)
+        {
+            return Apply(namespaces);
+        }
+
+        public ResolveGenericsSyntax(ResolveGenericsScope scope) : base(scope) { }
+    }
+
+    public class ResolveGenericsScope :
+            ScopeTransformation<ResolveGenericsExpression>
+    {
+        public ResolveGenericsScope() : base(new ResolveGenericsExpression()) { }
+    }
+
+    public class ResolveGenericsExpression :
+        ExpressionTransformation<ResolvedGenericsExpressionKind>
+    {
+        public ResolveGenericsExpression() : base(ex => new ResolvedGenericsExpressionKind(ex as ResolveGenericsExpression)) { }
+
+        public override TypedExpression Transform(TypedExpression ex)
+        {
+            ImmutableDictionary<QsTypeParameter, ResolvedType> types = ex.TypeParameterResolutions;
+
+            if (types.Any() && ex.Expression is QsExpressionKind<TypedExpression, Identifier, ResolvedType>.CallLikeExpression call)
+            {
+                string prefix = CreateTypePrefix(types);
+
+                // For now, only resolve identifiers of global callables
+                if (call.Item1.Expression is QsExpressionKind<TypedExpression, Identifier, ResolvedType>.Identifier id &&
+                    id.Item1 is Identifier.GlobalCallable globalCallable)
+                {
+                    // Rebuild the method with the updated identifier
+                    ex = new TypedExpression(
+                        QsExpressionKind<TypedExpression, Identifier, ResolvedType>.NewCallLikeExpression(
+                            new TypedExpression(
```

Comment:
For the sake of public documentation of what we discussed: 
We want to make this recursive since technically the left hand side of the call might be an partial application that resolves some of the type parameters etc. Hence you really wan to call this.Transform on the subexpression, and you want to pass the information on the type parameter resolution down. 
You should also think about how you want to handle the case where once you get to the identifier not all type parameters are resolved.

Reply:
This will need to be addressed later once there is a data structure for keeping tabs on information regarding generics.